common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,402 @@
+		encodedHistoryBatches, err := h.gcloudStorage.Get(ctx, URI, filename)
+
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+
+		if encodedHistoryBatches == nil {
+			return nil, &shared.InternalServiceError{Message: "Fail retrieving history file: " + URI.String() + "/" + filename}
+		}
+
+		batches, err := decodeHistoryBatches(encodedHistoryBatches)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		historyBatches = append(historyBatches, batches...)
+		historyBatches = historyBatches[token.NextBatchIdx:]
+
+		for _, batch := range historyBatches {
+			response.HistoryBatches = append(response.HistoryBatches, batch)
+			numOfBatches++
+			numOfEvents += len(batch.Events)
+
+		}
+
+		if numOfEvents >= request.PageSize {
+			break
+		}
+	}
+
+	if numOfBatches < len(response.HistoryBatches) {
+		token.NextBatchIdx += numOfBatches
+		nextToken, err := serializeToken(token)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		response.NextPageToken = nextToken
+	}
+
+	return response, nil
+}
+
+// ValidateURI is used to define what a valid URI for an implementation is.
+func (h *historyArchiver) ValidateURI(URI archiver.URI) (err error) {
+
+	if URI.Scheme() != URIScheme {
+		return archiver.ErrURISchemeMismatch
+	}
+
+	if URI.Path() == "" || URI.Hostname() == "" {
+		return archiver.ErrInvalidURI
+	}
+
+	return
+}
+
+func (h *historyArchiver) validateURI(ctx context.Context, URI archiver.URI) (err error) {
+	if err = h.ValidateURI(URI); err == nil {
+		_, err = h.gcloudStorage.Exist(ctx, URI, "")
+	}
+
+	return
+}
+
+func getNextHistoryBlob(ctx context.Context, historyIterator archiver.HistoryIterator) (*archiver.HistoryBlob, error) {
+	historyBlob, err := historyIterator.Next()
+	op := func() error {
+		historyBlob, err = historyIterator.Next()
+		return err
+	}
+	for err != nil {
+		if !common.IsPersistenceTransientError(err) {
+			return nil, err
+		}
+		if contextExpired(ctx) {
+			return nil, archiver.ErrContextTimeout
+		}
+		err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+	}
+	return historyBlob, nil
+}
+
+func historyMutated(request *archiver.ArchiveHistoryRequest, historyBatches []*shared.History, isLast bool) bool {
+	lastBatch := historyBatches[len(historyBatches)-1].Events
+	lastEvent := lastBatch[len(lastBatch)-1]
+	lastFailoverVersion := lastEvent.GetVersion()
+	if lastFailoverVersion > request.CloseFailoverVersion {
+		return true
+	}
+
+	if !isLast {
+		return false
+	}
+	lastEventID := lastEvent.GetEventId()
+	return lastFailoverVersion != request.CloseFailoverVersion || lastEventID+1 != request.NextEventID
+}
+
+func (h *historyArchiver) getHighestVersion(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*int64, *int64, *int64, error) {
```

Comment:
User can specify a workflow version in the request. But seems like this function will always return the highest version.

Reply:
Done. Also a new unittest was created: `TestGet_Success_UseProvidedVersion`